include/revng/TupleTree/DiffError.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/TupleTree/TupleTreePath.h"
+
+namespace model {
+
+class Binary;
+}
+namespace revng {
+
+class LocationBase {
+public:
+  virtual std::string toString() const = 0;
+  virtual ~LocationBase() = default;
+};
+
+template<typename RootType>
+class DiffLocation : public LocationBase {
+private:
+  TupleTreePath Path;
+
+public:
+  DiffLocation(TupleTreePath Path) : Path(Path) {}
+
+  std::string toString() const override {
+    auto AsString = pathAsString<RootType>(Path);
+    revng_assert(AsString.has_value());
+    return *AsString;
+  }
+
+  ~DiffLocation() override = default;
+};
+
+class RevngErrorBase : public llvm::ErrorInfo<RevngErrorBase> {
+public:
+  inline static char ID = '0';
+
+  using iterator = llvm::SmallVector<std::string, 4>::iterator;
+  using const_iterator = llvm::SmallVector<std::string, 4>::const_iterator;
+
+  using IDContainer = llvm::SmallVector<const void *, 4>;
+
+protected:
+  llvm::SmallVector<std::string, 4> Reasons;
+  llvm::SmallVector<std::unique_ptr<LocationBase>, 4> Locations;
+
+  IDContainer Ids;
+
+public:
+  explicit RevngErrorBase(const IDContainer &Ids) :
+    Reasons(), Ids(std::move(Ids)) {}
+  RevngErrorBase() = default;
+
+  static llvm::Error concatenate(llvm::Error SourceError,
+                                 llvm::StringRef NewReason,
+                                 std::unique_ptr<LocationBase> Loc) {
+
+    if (not SourceError.isA<RevngErrorBase>())
+      return SourceError;
+
+    llvm::Error ToReturn = llvm::Error::success();
+    return llvm::handleErrors(std::move(SourceError),
+                              [&](std::unique_ptr<revng::RevngErrorBase>
+                                    RevngError) {
+                                RevngError->Reasons.push_back(NewReason.str());
+                                RevngError->Locations.push_back(std::move(Loc));
+                                return llvm::Error(std::move(RevngError));
+                              });
+  }
+
+  // Returns the class ID for this type.
+  static const void *classID() { return &ID; }
+
+  iterator begin() { return Reasons.begin(); }
+
+  iterator end() { return Reasons.end(); }
+
+  [[nodiscard]] const_iterator begin() const { return Reasons.begin(); }
+
+  [[nodiscard]] const_iterator end() const { return Reasons.end(); }
+
+  size_t size() const { return Reasons.size(); }
+
+  const std::string &getMessage(size_t I) const { return Reasons[I]; }
+  std::string getLocation(size_t I) const { return Locations[I]->toString(); }
+
```

Comment:
Can we have the [two string discriminators](https://pad.rev.ng/vK82o4YtRXKv2A_ctMjHKg#) of the error type?
```c++
  virtual string getLocationTypeName();
  virtual string getErrorTypeName();
```